File: candidates/python/aiohttp-bridge/src/app/store.py
from __future__ import annotations

import asyncio
from app.models import TelemetryEnvelope


class InMemoryTelemetryStore:
    def __init__(self):
        self._lock = asyncio.Lock()
        self.drone_models: dict[str, str] = {}
        self.latest_telemetry: dict[str, TelemetryEnvelope] = {}

    async def register_drone(self, drone_id: str, model: str) -> None:
        async with self._lock:
            self.drone_models[drone_id] = model

    async def record_telemetry(self, telemetry: TelemetryEnvelope) -> None:
        async with self._lock:
            self.latest_telemetry[telemetry.drone_id] = telemetry
            if telemetry.drone_id not in self.drone_models:
                self.drone_models[telemetry.drone_id] = ""

    async def list_active_drones(self) -> list[str]:
        async with self._lock:
            return sorted(list(self.drone_models.keys()))

    async def latest_telemetry_for_drone(self, drone_id: str) -> TelemetryEnvelope | None:
        async with self._lock:
            return self.latest_telemetry.get(drone_id)

File: drone-gcs-bench/candidates/typescript/elysia-gateway/src/store.ts
export interface TelemetryPayload {
  lat: number;
  lon: number;
  alt: number;
  roll: number;
  pitch: number;
  yaw: number;
  battery: number;
  mode: string;
}

export interface TelemetryEnvelope {
  run_id: string;
  drone_id: string;
  seq: number;
  timestamp: number;
  payload: TelemetryPayload;
}

export class InMemoryTelemetryStore {
  private droneModels = new Map<string, string>();
  private latestTelemetryMap = new Map<string, TelemetryEnvelope>();

  public async registerDrone(droneId: string, model: string): Promise<void> {
    this.droneModels.set(droneId, model);
  }

  public async recordTelemetry(telemetry: TelemetryEnvelope): Promise<void> {
    this.latestTelemetryMap.set(telemetry.drone_id, telemetry);
    if (!this.droneModels.has(telemetry.drone_id)) {
      this.droneModels.set(telemetry.drone_id, "");
    }
  }

  public async listActiveDrones(): Promise<string[]> {
    return Array.from(this.droneModels.keys()).sort();
  }

  public async latestTelemetry(droneId: string): Promise<TelemetryEnvelope | null> {
    return this.latestTelemetryMap.get(droneId) || null;
  }
}
